### PR TITLE
Add function to fetch instance id

### PR DESCRIPTION
**Pull request by ncrothe**
Add convenience function erlcloud_aws:get_instance_id/1 to fetch instance id of current instance from AWS metadata

### CODE REVIEW

**Comment by ncrothe** [comment]
Not sure if erlcloud_aws is the right place, feel free to suggest other homes for the function.

**Comment by kkuzmin** [comment]
I vote for moving it to the new module named like `erlcloud_instance_metadata.erl` or similar.

**Comment by smilerlee** [comment]
Vote for `erlcloud_ec2_util.erl`

**Comment by ncrothe** [comment]
Maybe `erlcloud_meta.erl` and bundle this and any future functions exploiting the metadata endpoint there? Currently only some iam/credential metadata is being used (and from `erlcloud_aws.erl` which is why I ended up in there).

**Comment by motobob** [comment]
util.erl /meta.erl is preferable for such a minor auxiliary fun in spite it fully belongs to EC2 module. Probably RDS instance are exception here but I can't see Erlang app running on it.

However I would vote for more generalisation and having smth like
```erlang
get_instance_metadata() ->
   get_metadata(erlcloud_aws:default_config());
get_instance_metadata(Config) ->
   get_instace_metadata("", Config) .
get_instance_metadata(ItemPath, Config) ->
     Path = "http://169.254.169.254/latest/meta-data/" ++ ItemPath,
    %% your code goes here.
```

this way erlcloud_aws:get_credentials_from_metadata(Config) could be also simplified.

**Comment by ncrothe** [comment]
Moved the function into a new module erlcloud_ec2_meta and added new functionality to erlcloud_as dealing with lifecycle hooks. For now I implemented DescribeLifecycleHooks and CompleteLifecycleActions.

And sorry for the commit mess, I don't use git on a daily basis.

**Comment by waisbrot** [comment]
> And sorry for the commit mess, I don't use git on a daily basis.

At my office we've been using the new "Squash and Merge" button on GitHub and I think it's wonderful. We push the commit history, mess and all, and it can be useful to see the development process (like here, we'd know to help ncrothe brush up on Git (-;  ) but when everything's done and the PR is merged we get a lovely commit history on 'master'. (Nicer, IMO, without the merge-commits)

File: src/erlcloud_ec2_meta.erl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+get_instance_user_data() ->
+   get_instance_user_data(erlcloud_aws:default_config()).
+
+%%%---------------------------------------------------------------------------
+-spec get_instance_user_data( Config :: aws_config() ) -> {ok, binary()} | {error, string()}.
+%%%---------------------------------------------------------------------------
+%% @doc Retrieve the instance id for the instance this code is running on. Will fail if not an EC2 instance.
+%%
+%% This convenience function will retrieve the instance id from the AWS metadata available at 
+%% http://169.254.169.254/latest/meta-data/instance-id.
```
Comment (motobob) [inline src/erlcloud_ec2_meta.erl]: copy-paste error

File: src/erlcloud_ec2_meta.erl [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+   get_instance_dynamic_data("", Config).
+
+
+%%%---------------------------------------------------------------------------
+-spec get_instance_dynamic_data( ItemPath :: string(), Config :: aws_config() ) -> {ok, term()} | {error, string()}.
+%%%---------------------------------------------------------------------------
+%% @doc Retrieve the instance id for the instance this code is running on. Will fail if not an EC2 instance.
+%%
+%% This convenience function will retrieve the instance id from the AWS metadata available at 
+%% http://169.254.169.254/latest/meta-data/instance-id.
```
Comment (motobob) [inline src/erlcloud_ec2_meta.erl]: copy-paster error

File: src/erlcloud_as.erl [hunk 3]
```diff
@@ -412,6 +421,62 @@ detach_instances(InstanceIds, GroupName, ShouldDecrementDesiredCapacity, Config)
...
+describe_lifecycle_hooks(GroupName, Config) ->
+    describe_lifecycle_hooks(GroupName, [], Config).
+
+-spec describe_lifecycle_hooks(string(), list(string()), aws_config()) -> {ok, list(aws_autoscaling_lifecycle_hook())} | {error, term()}.
+describe_lifecycle_hooks(GroupName, LifecycleHookNames, Config) ->
+    Params = [{"AutoScalingGroupName", GroupName} | member_params("LifecycleHookNames.member.", LifecycleHookNames)],
+    case as_query(Config, "DescribeLifecycleHooks", Params, ?API_VERSION) of
+        {ok, Doc} ->
+            % XmlDoc = lists:flatten(xmerl:export_simple([Doc], xmerl_xml)),
+            % io:format("reverse ~n~p~n",[XmlDoc]),
```
Comment (motobob) [inline src/erlcloud_as.erl]: please remove these commented out lines

File: src/erlcloud_as.erl [hunk 4]
```diff
@@ -412,6 +421,62 @@ detach_instances(InstanceIds, GroupName, ShouldDecrementDesiredCapacity, Config)
...
+          lifecycle_transition = erlcloud_xml:get_text("LifecycleTransition", H)}.
+
+
+-spec complete_lifecycle_action(string(), string(), string(), string() | binary()) -> {ok, string()} | {error, term()}.
+complete_lifecycle_action(GroupName, LifecycleActionResult, LifecycleHookName, InstanceIdOrLifecycleActionToken) ->
+    complete_lifecycle_action(GroupName, LifecycleActionResult, LifecycleHookName, InstanceIdOrLifecycleActionToken, erlcloud_aws:default_config()).
+
+-spec complete_lifecycle_action(string(), string(), string(), string() | binary(), aws_config()) -> {ok, term()} | {error, term()}.
+complete_lifecycle_action(GroupName, LifecycleActionResult, LifecycleHookName, InstanceIdOrLifecycleActionToken, Config) ->
+    InstanceIdOrLifecycleActionTokenParam = case InstanceIdOrLifecycleActionToken of
```
Comment (motobob) [inline src/erlcloud_as.erl]: this is strange API.
can it be both? 
re: http://docs.aws.amazon.com/AutoScaling/latest/APIReference/API_CompleteLifecycleAction.html

**Comment by ncrothe** [comment]
Thanks for the feedback will get right to it.
Yes, the API is a bit strange. You can give either the token (which is generated by autoscaling on every lifecycle event and sent to SQS or SNS for processing); or you use the instance-id (if you know it). As far as I understand it is either one or the other. Compare http://docs.aws.amazon.com/autoscaling/latest/userguide/lifecycle-hooks.html (at the very bottom). I have implemented the API in erlcloud to accept either with the implicit assumption that string() is InstanceId and binary() is token.

In the use case I built this for, I use the instance id variant and call the complete-lifecycle-action when my instance logic is 100% up and running and ready to be put into rotation.

**Comment by ncrothe** [comment]
A Co-Worker raised the interesting point of erlcloud_ec2_meta fetching meta data from
http://169.254.169.254/**latest**/meta-data instead of say http://169.254.169.254/**2015-10-20**/meta-data , i.e. implying that a version step might break things.
The original token logic used latest, but what are everyone's current thoughts on this?

**Comment by ncrothe** [comment]
Forget that, found out that on eu-west-1a that particular meta-data version (**2015-10-20**) is not available, the newest there is **2014-11-05** and somewhere else I found the newer **2016-04-19**.
Now I'm completely at a loss and might lean towards accepting potential problems with **latest** vs. real problems with specific version numbering. Have contacted AWS support for clarification.

**Comment by ncrothe** [comment]
Not entirely surprising, Amazon said they are trying to ensure backwards compatibility, but if you need a specific version use the corresponding version path. http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-data-categories has the paths and corresponding version. Considering the non-latest metadata is IMHO a fringe case, I vote for leaving it at /latest and whoever needs something specific should patch his version.

**Comment by motobob** [comment]
let leave it `/latest`

**Comment by motobob** [comment]
@ncrothe please squash the history (git rebase -i HEAD~16) and fix the API for `InstanceIdOrLifecycleActionToken` list or binary so that get to be merged

**Comment by ncrothe** [comment]
Managed to squash my changes, but seem to have pulled in a ton of different changes via a clumsy rebase on an interlaced history. Any suggestions how to get rid of these?

Also changed the complete_lifecycle_action function to take a tagged tuple {instance_id | token, string()} to explicitly mark the parameter as either an instance_id or a lifecycle token.

**Comment by motobob** [comment]
@ncrothe 
git remote add upstream https://github.com/erlcloud/erlcloud
git fetch upstream
git rebase -i upstream/master
<pick only your commit> 
<if several use `s XXXX`, where XXXX is short commit>
<resolve any conflicts>
git push -f origin master

**Comment by ncrothe** [comment]
Thanks

File: src/erlcloud_ec2_meta.erl [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+%% @doc Retrieve the instance meta data for the instance this code is running on. Will fail if not an EC2 instance.
+%%
+%% This convenience function will retrieve the instance id from the AWS metadata available at 
+%% http://169.254.169.254/latest/meta-data/*
+%% ItemPath allows fetching specific pieces of metadata.
+%%
+%%
+get_instance_metadata(ItemPath, Config) ->
+    MetaDataPath = "http://169.254.169.254/latest/meta-data/" ++ ItemPath,
+    case erlcloud_aws:http_body(erlcloud_httpc:request(MetaDataPath, get, [], <<>>, erlcloud_aws:get_timeout(Config), Config)) of
```
Comment (motobob) [inline src/erlcloud_ec2_meta.erl]: this `case...of` is rudimentary

File: src/erlcloud_ec2_meta.erl [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+%%%---------------------------------------------------------------------------
+%% @doc Retrieve the user data for the instance this code is running on. Will fail if not an EC2 instance.
+%%
+%% This convenience function will retrieve the user data the instance was started with, i.e. what's available at 
+%% http://169.254.169.254/latest/user-data
+%%
+%%
+get_instance_user_data(Config) ->
+    UserDataPath = "http://169.254.169.254/latest/user-data/",
+    case erlcloud_aws:http_body(erlcloud_httpc:request(UserDataPath, get, [], <<>>, erlcloud_aws:get_timeout(Config), Config)) of
```
Comment (motobob) [inline src/erlcloud_ec2_meta.erl]: this case...of is rudimentary

File: src/erlcloud_ec2_meta.erl [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+   get_instance_dynamic_data("", Config).
+
+
+%%%---------------------------------------------------------------------------
+-spec get_instance_dynamic_data( ItemPath :: string(), Config :: aws_config() ) -> {ok, term()} | {error, string()}.
+%%%---------------------------------------------------------------------------
+
+get_instance_dynamic_data(ItemPath, Config) ->
+    DynamicDataPath = "http://169.254.169.254/latest/dynamic/" ++ ItemPath,
+    case erlcloud_aws:http_body(erlcloud_httpc:request(DynamicDataPath, get, [], <<>>, erlcloud_aws:get_timeout(Config), Config)) of
```
Comment (motobob) [inline src/erlcloud_ec2_meta.erl]: this case...of is rudimentary

**Comment by ncrothe** [comment]
Fair remark, have removed it and readded the fixes to src/erlcloud_aws.erl (must have gotten lost in the rebase somewhere).

File: src/erlcloud_as.erl [hunk 8]
```diff
@@ -412,6 +421,60 @@ detach_instances(InstanceIds, GroupName, ShouldDecrementDesiredCapacity, Config)
...
+          heartbeat_timeout = erlcloud_xml:get_integer("HeartbeatTimeout", H),
+          default_result = erlcloud_xml:get_text("DefaultResult", H),
+          lifecycle_transition = erlcloud_xml:get_text("LifecycleTransition", H)}.
+
+
+-spec complete_lifecycle_action(string(), string(), string(), {instance_id | token, string()}) -> {ok, string()} | {error, term()}.
+complete_lifecycle_action(GroupName, LifecycleActionResult, LifecycleHookName, InstanceIdOrLifecycleActionToken) ->
+    complete_lifecycle_action(GroupName, LifecycleActionResult, LifecycleHookName, InstanceIdOrLifecycleActionToken, erlcloud_aws:default_config()).
+
+-spec complete_lifecycle_action(string(), string(), string(), {instance_id | token, string()}, aws_config()) -> {ok, term()} | {error, term()}.
```
Comment (kkuzmin) [inline src/erlcloud_as.erl]: Suppose the return spec should be the same `{ok, string()} | {error, term()}` as for the `complete_lifecycle_action()` above

File: src/erlcloud_ec2_meta.erl [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+get_instance_metadata() ->
+   get_instance_metadata(erlcloud_aws:default_config()).
+
+-spec get_instance_metadata(Config :: aws_config() ) -> {ok, string()} | {error, string()}.
+get_instance_metadata(Config) ->
+   get_instance_metadata("", Config).
+
+
+%%%---------------------------------------------------------------------------
+-spec get_instance_metadata( ItemPath :: string(), Config :: aws_config() ) -> {ok, binary()} | {error, string()}.
```
Comment (kkuzmin) [inline src/erlcloud_ec2_meta.erl]: `erlcloud_aws:http_body()` has a bit wider return spec `{ok, binary()} | {error, tuple()}`.
Same below.

**Comment by kkuzmin** [comment]
+1 after a few minor spec checks.
Thank you, Christian

**Comment by ncrothe** [comment]
Good spot @kkuzmin. Fixed the specs and am starting to get the hang of git rebase. Maybe next time I'll even manage to use a branch.

**Pull request merged**